qubes/app.py

```diff
@@ -102,14 +102,19 @@ def wrapper(*args, **kwargs):
 class VirConnectWrapper:
     # pylint: disable=too-few-public-methods
 
-    def __init__(self, uri):
+    def __init__(self, uri, reconnect_cb=None):
         self._conn = libvirt.open(uri)
+        self._reconnect_cb = reconnect_cb
 
     def _reconnect_if_dead(self):
         is_dead = not self._conn.isAlive()
         if is_dead:
-            self._conn = libvirt.open(self._conn.getURI())
-            # TODO: re-register event handlers
+            uri = self._conn.getURI()
+            old_conn = self._conn
+            self._conn = libvirt.open(uri)
+            if callable(self._reconnect_cb):
+                self._reconnect_cb(old_conn)
+            old_conn.close()
```

Comment:
This causes regression: domain objects do have references to this connection too and closing it breaks `isAlive()` call, breaking reconnect logic. Keeping old connection alive is wrong, I'll adjust reconnection logic for domain objects.